src/codebuild-handler/index.ts

```diff
@@ -0,0 +1,86 @@
+import * as AWSLambda from 'aws-lambda';
+import * as AWS from 'aws-sdk';
+
+const cb = new AWS.CodeBuild();
+
+export async function onEventhandler(event: AWSLambda.CloudFormationCustomResourceEvent) {
+  switch (event.RequestType) {
+    case 'Create':
+    case 'Update':
+      return startBuild(event);
+    case 'Delete':
+      return onDelete();
+  }
+}
+
+export async function startBuild(event: AWSLambda.CloudFormationCustomResourceEvent) {
+  const requestType = event.RequestType;
+  const projectName = event.ResourceProperties.projectName;
+  if (requestType === 'Create' || requestType === 'Update') {
+    const buildOutput = await cb.startBuild({
+      projectName: projectName,
+    }).promise();
+
+    const buildId = buildOutput.build?.id;
+
+    if (buildId) {
+      return {
+        BuildId: buildId,
+      };
+    } else {
+      throw new Error('BuildId does not exist after CodeBuild:StartBuild call'); // pass to isComplete
+    }
+  }
+  if (requestType === 'Delete') {
+    return {};
+  }
+  throw new Error(`Invalid request type: ${requestType}`);
+}
+
+export async function onDelete() {
+  return {
+    BuildId: 'onDelete',
+  };
+}
+
+export async function isCompleteHandler(event: AWSLambda.CloudFormationCustomResourceEvent) {
+
+  const buildId = (event as any).BuildId; // BuildId is passed in from onEvent CodeBuild:StartBuild call
+
+  if (buildId === undefined) {
+    throw new Error('BuildId was not found or undefined');
+  }
+
+  if (buildId === 'onDelete') {
+    return { IsComplete: true };
+  }
+
+  const build = await cb.batchGetBuilds({
+    ids: [buildId],
+  }).promise();
+
+  // we should always have a build since we have a valid buildId
+  if (build.builds && build.builds.length > 0) {
```

Comment:
💅 I generally prefer early aborts, it makes it easier to reason about the happy path of code.

Instead of:

```ts
if (happyCondition) {
  // 
  // 
  // dozens of lines of code
  //
  //
} else {
  throw new Error('oh no');
}
```

Write:

```ts
if (!happyCondition) {
  throw new Error('oh no');
}

// ...
// 
// dozens of lines of code
//
//
```